src/wasm/wasm-type.cpp

```diff
@@ -45,300 +48,284 @@ struct TypeInfo {
+  }
+  WASM_UNREACHABLE("unexpected kind");
 }
 
-} // namespace std
-
-namespace wasm {
+bool HeapTypeInfo::operator==(const HeapTypeInfo& other) const {
+  if (kind != other.kind) {
+    return false;
+  }
+  switch (kind) {
+    case SignatureKind:
+      return signature == other.signature;
+    case StructKind:
+      return struct_ == other.struct_;
+    case ArrayKind:
+      return array == other.array;
+  }
+  WASM_UNREACHABLE("unexpected kind");
+}
 
-namespace {
+bool HeapTypeInfo::operator<(const HeapTypeInfo& other) const {
+  if (kind != other.kind) {
+    return kind < other.kind;
+  }
+  switch (kind) {
+    case SignatureKind:
+      return signature < other.signature;
+    case StructKind:
+      return struct_ < other.struct_;
+    case ArrayKind:
+      return array < other.array;
+  }
+  WASM_UNREACHABLE("unexpected kind");
+}
 
-struct TypeStore {
+template<typename Info> struct Store {
   std::mutex mutex;
 
-  // Track unique_ptrs for constructed types to avoid leaks
-  std::vector<std::unique_ptr<TypeInfo>> constructedTypes;
-
-  // Maps from constructed types to the canonical Type ID.
-  std::unordered_map<TypeInfo, uintptr_t> indices = {
-    // If a Type is constructed from a list of types, the list of types becomes
-    // implicitly converted to a TypeInfo before canonicalizing its id. This is
-    // also the case if a list of just one type is provided, even though such a
-    // list of types will be canonicalized to the BasicType of the single type.
-    // As such, the following entries are solely placeholders to enable the
-    // lookup of lists of just one type to the BasicType of the single type.
-    {TypeInfo(Tuple()), Type::none},
-    {TypeInfo({Type::unreachable}), Type::unreachable},
-    {TypeInfo({Type::i32}), Type::i32},
-    {TypeInfo({Type::i64}), Type::i64},
-    {TypeInfo({Type::f32}), Type::f32},
-    {TypeInfo({Type::f64}), Type::f64},
-    {TypeInfo({Type::v128}), Type::v128},
-    {TypeInfo({Type::funcref}), Type::funcref},
-    {TypeInfo(HeapType(HeapType::FuncKind), true), Type::funcref},
-    {TypeInfo({Type::externref}), Type::externref},
-    {TypeInfo(HeapType(HeapType::ExternKind), true), Type::externref},
-    {TypeInfo({Type::exnref}), Type::exnref},
-    {TypeInfo(HeapType(HeapType::ExnKind), true), Type::exnref},
-    {TypeInfo({Type::anyref}), Type::anyref},
-    {TypeInfo(HeapType(HeapType::AnyKind), true), Type::anyref},
-    {TypeInfo({Type::eqref}), Type::eqref},
-    {TypeInfo(HeapType(HeapType::EqKind), true), Type::eqref},
-    {TypeInfo({Type::i31ref}), Type::i31ref},
-    {TypeInfo(HeapType(HeapType::I31Kind), false), Type::i31ref},
-  };
+  // Track unique_ptrs for constructed types to avoid leaks.
+  std::vector<std::unique_ptr<Info>> constructedTypes;
 
-  uintptr_t canonicalize(const TypeInfo& info) {
+  // Maps from constructed types to their canonical Type IDs.
+  std::unordered_map<Info, uintptr_t> typeIDs;
+
+  typename Info::type_t canonicalize(Info info) {
```

Comment:
```suggestion
  typename Info::type_t canonicalize(const Info& info) {
```
Any reason for the change?

Reply:
There are two changes here:

1. I switched from having `canonicalize` returning an integer to having it return either `Type` or `HeapType` directly, since that's the only way it is ever used.

2. I switched from taking the `Info` (previously always a `TypeInfo`, now also possibly a `HeapTypeInfo`) by value to taking it be reference because `TypeInfo` can contain a `Tuple` and `HeapTypeInfo` can contain a `Struct`, both of which can have non-trivial copies.

Reply:
Sorry, but not sure how they are related to the `const Info&` thing..?

Reply:
Ah, sorry, I had confused myself 😓  Thanks for this catch; there was no reason for me to have stopped using const reference here.